src/js/abstract/base-component.js

```diff
@@ -0,0 +1,370 @@
+import getAttributes from '../get-attributes';
+import { publish, subscribe } from '../pubsub';
+import maybe from '../maybe';
+
+const THROWED_ERROR = 'throwed';
+const ids = {};
+const getId = (nodeName) => {
+  if (!(nodeName in ids)) {
+    ids[nodeName] = 0;
+  }
+
+  return ++ids[nodeName]; // eslint-disable-line no-plusplus
+};
+const lifecycleLogger = maybe((...args) => console.log(...args))()(true);
+
+/**
+ * Base class {BaseComponent}. This class checks if a template is set in the custom element
+ * and if yes appends it. It also appends custom styles to the top of the dom tree.
+ *
+ * **Light DOM**
+ * The light DOM are the provided children from the users of your component (light meaning easy to digest).
+ *
+ * ```html
+ * <axa-example>
+ *   <div>This is some light DOM for axa-example</div>
+ * </axa-example>
+ * ```
+ *
+ * **Local DOM**
+ * The local DOM is the DOM tree rendered by the component itself (in our case provided by `template`).
+ *
+ * ```js
+ * function(props, childrenFragment) {
+ *   return bel`<article>
+ *     ${childrenFragment} <!-- light DOM injection point -->
+ *   </article>`;
+ * }
+ * ```
+ *
+ * **Flattened DOM**
+ * The flattened DOM is the final product where the user's light DOM is injected into the Components local DOM.
+ *
+ * ```html
+ * <axa-example>
+ *   <article>
+ *     <div>This is some light DOM for axa-example</div> <!-- light DOM injection point -->
+ *   </article>
+ * </axa-example>
+ * ```
+ */
+export default class BaseComponent extends HTMLElement {
+  constructor(styles = '', template) {
+    super();
+
+    this._makeContextReady = this._makeContextReady.bind(this);
+    this._initialise(styles, template);
+    this._id = getId(this.nodeName);
+  }
+
+  /**
+   * _initialise - description
+   *
+   * @param  {type} styles description
+   * @param  {type} template description
+   * @return {type}        description
+   */
+  _initialise(styles, template = null) {
+    this._styles = styles;
+    this._template = template;
+    this._hasTemplate = !!template;
+    this._hasRendered = false;
+    this._isConnected = false;
+  }
+
+  /**
+   * connectedCallback - description
+   *
+   * @return {type}  description
+   */
+  connectedCallback() {
+    lifecycleLogger(this.logLifecycle)(`\n^^^ connectedCallback -> ${this.nodeName}#${this._id}`);
+
+    if (!this._isConnected) {
+      this.initialClassName = this.className;
+    }
+
+    this._appendStyles();
+    this.render();
```

Comment:
@LucaMele 
Since `render` is called synchronously, neither any components will have any chance to complete it's custom `connectedCallback` logic before `render` is called. (`willRender` and `didRender` callbacks mitigate this issue)